tsec-http4s/src/main/scala/tsec/authentication/JWKPublicKeyRSAAuthenticator.scala

```diff
@@ -0,0 +1,197 @@
+package tsec.authentication
+
+import java.security.KeyFactory
+import java.security.spec.RSAPublicKeySpec
+import java.time.Instant
+import java.util.concurrent.atomic.AtomicReference
+
+import cats.data.OptionT
+import cats.effect.{Effect, Sync}
+import cats.syntax.all._
+import fs2.Stream
+import fs2.async.Ref
+import io.circe.Decoder
+import io.circe.generic.auto._
+import io.circe.parser.decode
+import org.http4s.circe._
+import org.http4s.client.blaze._
+import org.http4s.headers.Authorization
+import org.http4s.{AuthScheme, Credentials, EntityDecoder, Request, Response, Uri}
+import tsec.common.{SecureRandomId, _}
+import tsec.jws.signature.{JWSSigCV, JWTSig}
+import tsec.signature.jca.{JCASigTag, SigPublicKey}
+
+import scala.concurrent.duration.FiniteDuration
+
+final case class AugmentedJWK[A, I](
+    id: SecureRandomId,
+    jwt: JWTSig[A],
+    identity: I,
+    expiry: Instant,
+    lastTouched: Option[Instant]
+)
+
+final case class Modulus(value: BigInt) extends AnyVal
+final case class Exponent(value: BigInt) extends AnyVal
+final case class JWK(kid: String, kty: String, use: String, n: Modulus, e: Exponent)
+
+final case class JWKS(keys: List[JWK])
+
+class KeyRegistry[F[_], A: JCASigTag](uri: Uri, minFetchDelay: FiniteDuration)(implicit E: Effect[F]) {
+
+  implicit val jwksDecoder: EntityDecoder[F, JWKS] = jsonOf[F, JWKS]
+  implicit val modulusDecoder: Decoder[Modulus]    = Decoder.decodeString.map(s => Modulus(BigInt(1, s.base64UrlBytes)))
+  implicit val exponentDecoder: Decoder[Exponent]  = Decoder.decodeString.map(s => Exponent(BigInt(1, s.base64UrlBytes)))
+
+  private val keys      = new AtomicReference(Map[String, SigPublicKey[A]]())
```

Comment:
Okay, so the proper, functional way to do this, would be to make `keyRegistry` take an `Ref[F, Map[String, SigPublicKey[A]]` and `lastFetch` a `Ref[F, Instant]`.

For the record, `lastFetch` is _never_ set to `none` after being retrieved for the first time, so there's no reason to have it in option at all. Simply making the first set time something like the beginning of unix epoch ensures that the first call will always fetch.

Instantiation of `KeyRegistry` as a class is then `F[KeyRegistry[F, A]]`, which is fine and totally normal to signal the fact that it allocates mutable references